### PR TITLE
WIP: Rover Sailboat Support

### PR DESCRIPTION
**Pull request by IamPete1**
First steps of adding sailing support, hope to work on this over the next few weeks and get sailboat support up and running, see https://github.com/ArduPilot/ardupilot/issues/7799

discussion here: https://discuss.ardupilot.org/t/sailboat-support/32060

### CODE REVIEW

**Comment by IamPete1** [comment]
https://github.com/ArduPilot/ardupilot/issues/7799

**Comment by rmackay9** [comment]
Ah this is interesting and looks a bit similar to a branch I've started working on as well: https://github.com/rmackay9/rmackay9-ardupilot/commits/sailboat1

I'm very happy to have someone else give this a go though 'cuz I have a lot on my plate.  Or perhaps we could do a bit of cooperative development..

File: libraries/SRV_Channel/SRV_Channel.h [hunk 1]
```diff
@@ -121,6 +121,7 @@ class SRV_Channel {
         k_dspoilerLeft2         = 86,           ///< differential spoiler 2 (left wing)
         k_dspoilerRight2        = 87,           ///< differential spoiler 2 (right wing)
         k_winch                 = 88,
+        k_main_sheet            = 89,
```
[the remark this answers is not in the document]
Comment (amilcarlucas) [reply libraries/SRV_Channel/SRV_Channel.h]: Or call it main_sail_sheet?

**Comment by IamPete1** [comment]
Cooperative development sounds great, I have merged in your work and done some more work on the Library,

File: libraries/AP_WindVane/AP_WindVane.cpp [hunk 2]
```diff
@@ -94,18 +121,18 @@ float AP_WindVane::get_apparent_wind_direction_rad()
...
             
         case WindVaneType::WINDVANE_ANALOG_PIN:
             apparent_angle = read_analog();
             break;
             
-        default : // Assume head to wind at home locaiton   
-            float bearing = home_bearing();
-            apparent_angle = fabsf(wrap_PI(ahrs.yaw - bearing)); // This is a aproximation as we are not considering boat speed and wind speed yet
+        default : // Assume head to wind at home locaiton, apparent wind angle must be caculated
+            float bearing = rover.ahrs.get_home().heading;
```
Comment (rmackay9) [inline libraries/AP_WindVane/AP_WindVane.cpp]: it's not possible to access the vehicle class object from within a library.

File: libraries/AP_WindVane/AP_WindVane.cpp [hunk 3]
```diff
@@ -128,19 +155,19 @@ float AP_WindVane::get_absolute_wind_direction_rad()
             return bearing;
 
         case WindVaneType::WINDVANE_NONE:
-            bearing = home_bearing();
+            bearing =  rover.ahrs.get_home().heading;
```
Comment (rmackay9) [inline libraries/AP_WindVane/AP_WindVane.cpp]: I don't think home has a heading..

**Comment by peterbarker** [comment]
On Sun, 19 Aug 2018, Randy Mackay wrote:

> +            float bearing = rover.ahrs.get_home().heading;
> 
> it's not possible to access the vehicle class object from within a
> library.

Try AP::ahrs().get_home() instead

File: libraries/AP_WindVane/AP_WindVane.h [hunk 4]
```diff
@@ -47,10 +47,10 @@ class AP_WindVane
     void update();
 
     // get the apparent wind direction in radians
-    float get_apparent_wind_direction_rad();
+    float get_apparent_wind_direction_rad(void);
```
Comment (rmackay9) [inline libraries/AP_WindVane/AP_WindVane.h]: I'm not a fan of adding the "void" actually.. there are various opinions in the group but..

File: libraries/AP_WindVane/AP_WindVane.h [hunk 5]
```diff
@@ -60,15 +60,15 @@ class AP_WindVane
     static AP_WindVane *_s_instance;
 
     // Wind Vane parameters
-    AP_Int8  _type;             // type of windvane being used
-    AP_Int8  _pin;              // analog or pwm pin connected to sensor
-    AP_Float _analog_volt_low;  // voltage when windvane is pointing at 0 degrees
-    AP_Float _analog_volt_high; // voltage when windvane is pointing at 359 degrees
-    AP_Int16 _pwm_low;          // PWM when 'windvane' is pointing at 0 degrees
-    AP_Int16 _pwm_value;        // PWM when 'windvane' is pointing at 359 degrees
+    AP_Int8 _WindVane_Type;             // type of windvane being used
```
Comment (rmackay9) [inline libraries/AP_WindVane/AP_WindVane.h]: no need to prepend _WindVane.  We are already inside the AP_WindVane class... also we normally don't use camel case but instead small letters with underscores..

**Comment by lvale** [comment]
Will you implement some kind of limit of list, and/or implement of main sail vs list vs course control?

**Comment by IamPete1** [comment]
Yeah, That's the plan, see the TODO list in sailboat.cpp, if you can think of another useful stuff to add that would be great, so long as it not too complicated ;)

**Comment by lvale** [comment]
On a brief overview I would have a larger angle for the main sail, with 0 being in line with the ship centerline and then +/- angle to be defined. On RC sailboats, sometimes you might want to go to more than 90 from centerline for some movements. Granted that it's not used frequently, and I usually use a switch on the radio to enable the extended range. The +/- distinction would be used for port/starboard distinction

**Comment by IamPete1** [comment]
got it compiling, for the 405-wing in any-case, my trouble was that I had forgotten to add the library to the makefiles. I think the rest should compile I swap sqrt for sqrtf i think. 

No luck with AP::ahrs().yaw though, possibly i'm using it wrong, any ideas?

**Comment by IamPete1** [comment]
It builds! should be ready for testing in manual and acro modes. 

It fails SITL tho, i'm not too sure why the logs seem inconclusive to me. 

I need some advice on the best way to implement a calibration method for the windvane, I guess a mavlink triggered calibration would be too complicated to set up so we could do a esc style calibration  triggered off a parameter change and a reboot I guess. Also I'm not too sure about the correct method for setting up a sailing logging type. 

In the mean time I will have a look for papers on to see if there is a nice method for controlling heading vs speed vs mainsheet vs heel, sort of like a sailing TECS. If there is it would make scene I think to start writing the auto code with that in mind even if we don't implement it straight away.

File: libraries/SRV_Channel/SRV_Channel.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (IamPete1) [reply libraries/SRV_Channel/SRV_Channel.h]: have done exactly that, its now k_mainsail_sheet
thanks for the review

**Comment by rmackay9** [comment]
@IamPete1, I've rebased on master, "squashed" some commits together, corrected some spelling mistakes and formatting issues and put in a branch in my repo: https://github.com/rmackay9/rmackay9-ardupilot/commits/iampete1-sailboat

Could you work off this branch if possible?

**Comment by rmackay9** [comment]
As a side note, I really like the idea of storing the home location at arming and using that as the wind direction.   I think it would be most intuitive if the user faced the vehicle into the wind when arming.. i'm not sure if it's implemented that way or if the user faces it downwind..

**Comment by IamPete1** [comment]
@rmackay9 thanks for the help, 
I will clone your branch and continue from there, or would it be better to do pull requests to your repo? I find this git stuff more confusing than the code. I think the heading when armed is correct, the wind direction is defined as the direction it is blowing from so it should be correct. I think next I will tackle the calibration code, I guess we could run the calibration in the library initialisation or when you try and arm, not sure which is best.

**Comment by rmackay9** [comment]
@IamPete1, I think you're doing the most work on the sailboat stuff and I'm just poking it now and then so for now at least, maybe it's best for the PRs to keep coming from you.

Re calibration, I was thinking that as a first step it could be a bit of a manual process for the user.  If we send the wind direction to the user using the [WIND](http://mavlink.org/messages/ardupilotmega#WIND) or [WIND_COV](http://mavlink.org/messages/common#WIND_COV) message then I think the user could see that info in the ground station (if using MP, I think it will show up in the Flight Data screen's status tab).  Then the user can poke away at some of the wind vane parameters to try and get it right.  This is just an idea though..

**Comment by rmackay9** [comment]
We should probably also add a new dataflash log message (maybe called SAIL?) that outputs the important bits of info.  That would probably be the apparent and absolute wind direction, mainsail_angle and/or the final mainsail value (i.e. value sent to g2.motors.set_mainsail()).  Anyway, I can potentially add this stuff at some point if you don't get to it.

**Comment by IamPete1** [comment]
@rmackay9  OK, Thanks, I assumed setting up the mavlink stuff would be more complex than the calibration, Ill have a look as setting up mavlink and logs first, would defiantly be nice to know the library is working as expected.

**Comment by rmackay9** [comment]
@IamPete1, I've resolved some merge issues related to another of my PRs going in and also added the mavlink reporting and dataflash logging in this branch: https://github.com/rmackay9/rmackay9-ardupilot/commits/iampete1-sailboat.

Hopefully I'm not re-doing stuff you've already done... I'm keen to test some time in the next week 'cuz my sailboat is nearly complete.

**Comment by IamPete1** [comment]
@rmackay9 Great, thanks Randy, I was just about to have a look at that stuff, I have been concentration on the rebuild of my boat since the code more or less ready for testing, hope to have it ready in the next day or two. I think the next thing in the list should be some sort of main-sheet heel control so it will sheet out instead of getting knocked down and then on to auto stuff, assuming what we have done so far works.

**Comment by IamPete1** [comment]
Closing this pull request and opening a new one from a new branch https://github.com/ArduPilot/ardupilot/pull/9322

File: libraries/AP_WindVane/AP_WindVane.h (hunk 4 above)
Comment (junwoo091400) [reply libraries/AP_WindVane/AP_WindVane.h]: @rmackay9 Hey, I was just curious whist reading about this PR. What is an advantage of not adding the 'void' explicitly? Is it more readable- or such?

**Pull request closed without merging**